src/context/UserContext.tsx

```diff
@@ -0,0 +1,103 @@
+import {
+  createContext,
+  useContext,
+  useReducer,
+  ReactNode,
+  Dispatch,
+} from 'react';
+
+interface UserState {
+  name: string;
+  email: string;
+  isLoggedIn: boolean;
+  currency: string;
+  language: string;
+  theme: string;
+  notificationSettings: {
+    expenseAlert: boolean;
+    budget: boolean;
+    tipsArticle: boolean;
+  };
+}
+
+type UserAction =
```

Comment:
I'd change this type a little bit.
1. Add enum for the type: 
``` 
 enum UserActionType {
  SET_USER = 'SET_USER',
  LOG_IN = 'LOG_IN',
  LOG_OUT = 'LOG_OUT',
  SET_LANGUAGE = 'SET_LANGUAGE',
  SET_THEME = 'SET_THEME',
  SET_CURRENCY = 'SET_CURRENCY',
  TOGGLE_NOTIFICATION = 'TOGGLE_NOTIFICATION',
}
```


2. Create type for each action type, then combine them using union. This will add a restriction to the data in the payload. E.g. if SET_LANGAUGE action type is used, the language in the payload MUST be provided.
```
type SetUserAction = { type: UserActionType.SET_USER; payload: Partial<UserState> };
type LogInAction = { type: UserActionType.LOG_IN };
type LogOutAction = { type: UserActionType.LOG_OUT };
type SetLanguageAction = { type: UserActionType.SET_LANGUAGE; payload: { language: string } };
type SetThemeAction = { type: UserActionType.SET_THEME; payload: { theme: string } };
type SetCurrencyAction = { type: UserActionType.SET_CURRENCY; payload: { currency: string } };
type ToggleNotificationAction = { type: UserActionType.TOGGLE_NOTIFICATION; payload: keyof NotificationSettings };

type UserAction =
  | SetUserAction
  | LogInAction
  | LogOutAction
  | SetLanguageAction
  | SetThemeAction
  | SetCurrencyAction
  | ToggleNotificationAction;
```

Reply:
Fixed. But honestly ı did not understand why we create type for each action type.